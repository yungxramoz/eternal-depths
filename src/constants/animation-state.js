const IDLE_STANDING = 'animate-idle-standing'
const IDLE_FLYING = 'animate-idle-flying'
const ATTACKING = 'animate-attacking'
const DAMAGING = 'animate-damaging'

export const ANIMATION_STATE = {
  IDLE_STANDING,
  IDLE_FLYING,
  ATTACKING,
  DAMAGING,
}
