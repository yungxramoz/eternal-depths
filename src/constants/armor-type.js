export const HELMET = Object.freeze({
  type: 'helmet',
  stats: {
    health: 0,
    precision: 0,
  },
})
export const ARMOR = Object.freeze({
  type: 'armor',
  stats: {
    health: 0,
    strength: 0,
    agility: 0,
  },
})
export const GLOVES = Object.freeze({
  type: 'gloves',
  stats: {
    strength: 0,
    agility: 0,
    precision: 0,
  },
})
export const GREAVES = Object.freeze({
  type: 'greaves',
  stats: {
    agility: 0,
    precision: 0,
  },
})

const ARMOR_TYPE = {
  HELMET,
  ARMOR,
  GLOVES,
  GREAVES,
}

export const ARMOR_TYPES = Object.values(ARMOR_TYPE)

export default ARMOR_TYPE
