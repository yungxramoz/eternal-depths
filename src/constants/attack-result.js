const HIT = 'hit'
const CRITICAL = 'critical'
const EVADED = 'evaded'
const MISSED = 'missed'

const ATTACK_RESULT = {
  HIT,
  CRITICAL,
  EVADED,
  MISSED,
}

export default ATTACK_RESULT
