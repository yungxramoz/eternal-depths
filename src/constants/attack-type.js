import STAT from './stat'

const NORMAL = 'normal'

export const ATTACK_TYPE = {
  NORMAL,
}

export const EMPTY_ATTACK = Object.freeze({
  name: 'Empty Attack',
  description: 'An empty attack.',
  icon: 'attack-empty',
  type: 'normal',
  selfHealAmount: 0,
  selfInflictedAmount: 0,
  hitCount: 1,
  additionalDamage: 0,
  damageDecrease: 0,
  criticalChance: 0,
  cooldown: 0,
  buffs: [],
})

export const BASE_ATTACK = Object.freeze({
  name: 'Base Attack',
  description: 'A basic attack that deals damage to the enemy.',
  icon: 'attack-base',
  type: 'normal',
  selfHealAmount: 0,
  selfInflictedAmount: 0,
  hitCount: 1,
  additionalDamage: 0,
  damageDecrease: 0,
  criticalChance: 0,
  cooldown: 0,
  buffs: [],
})

const DRAIN = Object.freeze({
  name: 'Drain',
  description: 'Steal health from the enemy.',
  icon: 'attack-drain',
  type: 'normal',
  selfHealAmount: 'auto',
  selfInflictedAmount: 0,
  hitCount: 1,
  additionalDamage: 0,
  criticalChance: 0,
  cooldown: 2,
  buffs: [],
})

const DANCE_OF_THE_DEAD = Object.freeze({
  name: 'Dance of the Dead',
  description: 'Attack the enemy multiple times and increase agility.',
  icon: 'attack-dance-of-the-dead',
  type: 'normal',
  selfHealAmount: 0,
  selfInflictedAmount: 0,
  hitCount: 3,
  additionalDamage: 0,
  criticalChance: 0,
  cooldown: 2,
  buffs: [
    {
      stat: STAT.AGILITY,
      value: 5,
      duration: 1,
    },
  ],
})

const BLOODLUST = Object.freeze({
  name: 'Bloodlust',
  description: 'Increase damage and critical chance.',
  icon: 'attack-bloodlust',
  type: 'normal',
  selfHealAmount: 0,
  selfInflictedAmount: 0,
  hitCount: 1,
  cooldown: 2,
  additionalDamage: 0,
  buffs: [
    {
      stat: STAT.STRENGTH,
      value: 5,
      duration: 1,
    },
    {
      stat: STAT.PRECISION,
      value: 5,
      duration: 1,
    },
  ],
})

const BERSERK_STRIKE = Object.freeze({
  name: 'Berserk Strike',
  description: 'Deal massive damage to the enemy and inflicts damage to self.',
  icon: 'attack-berserk-strike',
  type: 'normal',
  selfHealAmount: 0,
  selfInflictedAmount: 6,
  hitCount: 1,
  additionalDamage: 20,
  criticalChance: 0,
  cooldown: 3,
  buffs: [],
})

const DOUBLE_STRIKE = Object.freeze({
  name: 'Double Strike',
  description: 'Attack the enemy twice.',
  icon: 'attack-double-strike',
  type: 'normal',
  selfHealAmount: 0,
  selfInflictedAmount: 0,
  hitCount: 2,
  additionalDamage: 0,
  criticalChance: 0,
  cooldown: 1,
  buffs: [],
})

const DEATH_BLOW = Object.freeze({
  name: 'Death Blow',
  description: 'Deal massive damage to the enemy.',
  icon: 'attack-death-blow',
  type: 'normal',
  selfHealAmount: 0,
  selfInflictedAmount: 0,
  hitCount: 1,
  additionalDamage: 50,
  criticalChance: 0,
  cooldown: 5,
  buffs: [],
})

export const ATTACK = {
  DRAIN,
  DANCE_OF_THE_DEAD,
  BLOODLUST,
  BERSERK_STRIKE,
  DOUBLE_STRIKE,
  DEATH_BLOW,
}
