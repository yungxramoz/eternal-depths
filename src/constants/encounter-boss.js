export const FIRE_DRAGON = Object.freeze({
  name: 'Fire Dragon',
  fileName: 'fire-dragon.png',
  baseStats: {
    health: 3,
    strength: 3,
    agility: 2,
    precision: 2,
  },
  style: {
    width: '50vw',
    top: '0px',
    left: '0px',
  },
})

export const LICH = Object.freeze({
  name: 'Lich',
  fileName: 'lich.png',
  baseStats: {
    health: 2,
    strength: 2,
    agility: 3,
    precision: 3,
  },
  style: {
    width: '50vw',
    top: '0px',
    left: '0px',
  },
})

export const MAD_MAGE = Object.freeze({
  name: 'Mad Mage',
  fileName: 'mad-mage.png',
  baseStats: {
    health: 2,
    strength: 3,
    agility: 2,
    precision: 3,
  },
  style: {
    width: '50vw',
    top: '0px',
    left: '0px',
  },
})

export const SHADOW_DRAGON = Object.freeze({
  name: 'Shadow Dragon',
  fileName: 'shadow-dragon.png',
  baseStats: {
    health: 2,
    strength: 4,
    agility: 2,
    precision: 2,
  },
  style: {
    width: '50vw',
    top: '0px',
    left: '0px',
  },
})

export const VAMPIRE_LORD = Object.freeze({
  name: 'Vampire Lord',
  fileName: 'vampire-lord.png',
  baseStats: {
    health: 3,
    strength: 2,
    agility: 3,
    precision: 3,
  },
  style: {
    width: '50vw',
    top: '0px',
    left: '0px',
  },
})

export const BossEncounters = [
  FIRE_DRAGON,
  LICH,
  MAD_MAGE,
  SHADOW_DRAGON,
  VAMPIRE_LORD,
]
