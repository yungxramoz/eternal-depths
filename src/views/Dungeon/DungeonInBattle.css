.combat-control-container {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: flex-start;
    padding-top: 1rem;
    column-gap: 2rem;
}