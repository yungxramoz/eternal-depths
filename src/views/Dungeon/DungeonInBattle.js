import React, { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import AttackButton from '../../components/molecules/AtackButton/AttackButton'
import HpProgressBar from '../../components/molecules/ProgressBar/HpProgressBar'
import {
  attackEffects,
  calculatedCharacterStats,
  damageCharacter,
} from '../../store/character/characterSlice'
import {
  animateIdle,
  attack,
  battleDefeat,
  damageEncounter,
} from '../../store/game/gameSlice'
import { calculateDamage } from '../../utils/attack'
import './DungeonInBattle.css'
import { EMPTY_ATTACK } from '../../constants/attack-type'

const DungeonInBattle = ({ children }) => {
  const dispatch = useDispatch()
  const character = useSelector((state) => state.character.current)
  const encounter = useSelector((state) => state.game.encounter)
  const encounterTurn = useSelector((state) => state.game.encounterTurn)
  const characterStats = useSelector(calculatedCharacterStats)

  useEffect(() => {
    setTimeout(() => {
      if (encounterTurn) {
        if (encounter.hp > 0) {
          dispatch(attack())
          const damage = calculateDamage(
            EMPTY_ATTACK,
            encounter.stats,
            encounter.minDamage,
            encounter.maxDamage,
            characterStats,
          )
          dispatch(damageCharacter(damage))
        }
      } else {
        dispatch(animateIdle())
      }
    }, 500)
  }, [encounterTurn, dispatch, encounter, character.hp, characterStats])

  useEffect(() => {
    if (character.hp <= 0) {
      dispatch(battleDefeat())
    }
  }, [character.hp, dispatch])

  const invokeAttack = (attack) => {
    const { minDamage, maxDamage } = character.items.weapon.stats
    const damage = calculateDamage(
      attack,
      characterStats,
      minDamage,
      maxDamage,
      encounter.stats,
    )
    dispatch(damageEncounter(damage))
    dispatch(attackEffects({ attack, dealtDamage: damage }))
  }

  return (
    <>
      {children}
      <HpProgressBar
        prefix="Your "
        currentHp={character.hp}
        maxHp={character.maxHp}
      />
      <div className="combat-control-container">
        {character.attacks.map((attack) => (
          <AttackButton
            key={attack.id}
            disabled={attack.currentCooldown > 0 || encounterTurn}
            attack={attack}
            text={attack.name}
            onClick={() => invokeAttack(attack)}
          ></AttackButton>
        ))}
      </div>
    </>
  )
}

export default DungeonInBattle
