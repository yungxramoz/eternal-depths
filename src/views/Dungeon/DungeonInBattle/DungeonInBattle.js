import React, { useEffect, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import DamageIndicator from '../../../components/atoms/DamageIndicator/DamageIndicator'
import AttackButton from '../../../components/molecules/AtackButton/AttackButton'
import Encounter from '../../../components/molecules/Encounter/Encounter'
import HpProgressBar from '../../../components/molecules/ProgressBar/HpProgressBar'
import { EMPTY_ATTACK } from '../../../constants/attack-type'
import {
  calculatedCharacterStats,
  characterAnimateIdle,
  characterAttackEffects,
  characterDamage,
  characterMaxHp,
  encounterAnimateIdle,
  encounterAttack,
  encounterDamage,
  updateGameCycleState,
} from '../../../store/game/gameSlice'
import { calculateDamage } from '../../../utils/attack'
import './DungeonInBattle.css'

const DungeonInBattle = () => {
  const dispatch = useDispatch()
  const character = useSelector((state) => state.game.character.current)
  const encounter = useSelector((state) => state.game.encounter.current)
  const isEncounterTurn = useSelector((state) => state.game.isEncounterTurn)
  const turnCounter = useSelector((state) => state.game.turnCounter)
  const encounterAnimationState = useSelector(
    (state) => state.game.encounter.animation,
  )
  const characterStats = useSelector(calculatedCharacterStats)
  const charMaxHp = useSelector(characterMaxHp)

  const [isDisabled, setIsDisabled] = useState(false)

  const [messages, setMessages] = useState([])

  const addIndicator = (type, damage) => {
    setMessages((prev) => [...prev, { type, damage }])
  }

  useEffect(() => {
    dispatch(updateGameCycleState())
    if (isEncounterTurn && encounter.hp > 0) {
      invokeEncounterAttack()
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [turnCounter])

  const invokeAttack = async (attack) => {
    setIsDisabled(true)
    let dealtDamage = 0
    const hitCount = attack.hitCount
    for (let i = 0; i < hitCount; i++) {
      const { damage, result } = calculateDamage(
        attack,
        characterStats,
        encounter.stats,
      )
      addIndicator(result, damage)
      dispatch(encounterDamage(damage))
      dealtDamage += damage
      await new Promise((resolve) => setTimeout(resolve, 250))
      dispatch(encounterAnimateIdle())
      await new Promise((resolve) => setTimeout(resolve, 50))
    }
    await dispatch(characterAttackEffects({ attack, dealtDamage }))
    setIsDisabled(false)
  }

  const invokeEncounterAttack = async () => {
    setIsDisabled(true)
    const stats = {
      ...encounter.stats,
      minDamage: encounter.minDamage,
      maxDamage: encounter.maxDamage,
    }
    const { damage } = calculateDamage(
      EMPTY_ATTACK,
      stats,
      characterStats,
      true,
    )
    dispatch(encounterAttack())
    dispatch(characterDamage(damage))
    await new Promise((resolve) => setTimeout(resolve, 250))
    dispatch(encounterAnimateIdle())
    dispatch(characterAnimateIdle())
    await new Promise((resolve) => setTimeout(resolve, 50))
    setIsDisabled(false)
  }

  return (
    <>
      <DamageIndicator messages={messages} setMessages={setMessages} />
      <Encounter
        encounterAnimationState={encounterAnimationState}
        encounter={encounter}
        showHp
      />
      <HpProgressBar
        prefix="Your "
        currentHp={character.hp}
        maxHp={charMaxHp}
      />
      <div className="combat-control-container">
        {character.attacks.map((attack) => (
          <AttackButton
            key={attack.id}
            disabled={
              attack.currentCooldown > 0 || isEncounterTurn || isDisabled
            }
            attack={attack}
            text={attack.name}
            onClick={() => invokeAttack(attack)}
          ></AttackButton>
        ))}
      </div>
    </>
  )
}

export default DungeonInBattle
