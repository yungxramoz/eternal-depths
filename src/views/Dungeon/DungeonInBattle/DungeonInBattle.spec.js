import { configureStore } from '@reduxjs/toolkit'
import { fireEvent, render, screen } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import gameReducer, { damageEncounter } from '../../../store/game/gameSlice'
import DungeonInBattle from './DungeonInBattle'
import { BASE_ATTACK } from '../../../constants/attack-type'

describe('DungeonInBattle', () => {
  const initialState = {
    current: {
      look: 1,
      name: '',
      level: 1,
      xp: 0,
      stats: {
        health: 1,
        strength: 1,
        agility: 1,
        precision: 1,
      },
      items: {
        helmet: null,
        armor: null,
        weapon: {
          name: 'Rusty Sword',
          stats: {
            minDamage: 1,
            maxDamage: 3,
          },
        },
        shield: null,
        greaves: null,
      },
      attacks: [
        {
          ...BASE_ATTACK,
          id: 1,
          currentCooldown: 0,
        },
      ],
    },
    availableAttributePoints: 2,
  }
  let store
  beforeEach(() => {
    store = configureStore({
      reducer: {
        game: gameReducer,
      },
      preloadedState: {
        character: {
          ...initialState,
          current: {
            ...initialState.current,
            hp: 100,
            maxHp: 100,
            attacks: [
              {
                id: 1,
                name: 'Attack 1',
                currentCooldown: 0,
                fileName: 'mockAttack.png',
              },
            ],
            stats: { health: 1, strength: 1, agility: 1, precision: 1 },
          },
        },
        game: {
          isEncounterTurn: false,
          encounter: {
            hp: 100,
            maxHp: 100,
            minDamage: 1,
            maxDamage: 3,
            stats: { health: 1, strength: 1, agility: 1, precision: 1 },
          },
        },
      },
    })
  })
  it('renders correctly', () => {
    render(
      <Provider store={store}>
        <DungeonInBattle />
      </Provider>,
    )
    expect(document.body).toMatchSnapshot('DungeonInBattle')
  })
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('dispatches damageEncounter and attackEffects actions when an attack button is clicked', () => {
    store.dispatch = jest.fn()
    render(
      <Provider store={store}>
        <DungeonInBattle />
      </Provider>,
    )
    fireEvent.click(screen.getByRole('button', { name: 'Attack 1' }))
    expect(store.dispatch).toHaveBeenCalledTimes(2)
    expect(store.dispatch).toHaveBeenCalledWith(
      expect.objectContaining({
        type: 'game/damageEncounter',
        payload: expect.any(Number),
      }),
    )
    expect(store.dispatch).toHaveBeenCalledWith(
      expect.objectContaining({
        type: 'character/attackEffects',
        payload: {
          attack: expect.any(Object),
          dealtDamage: expect.any(Number),
        },
      }),
    )
  })
  it('dispatches animateIdle action when isEncounterTurn is false', async () => {
    jest.useFakeTimers()
    store.dispatch = jest.fn()
    render(
      <Provider store={store}>
        <DungeonInBattle />
      </Provider>,
    )
    jest.runAllTimers()
    expect(store.dispatch).toHaveBeenCalledWith(
      expect.objectContaining({ type: 'game/animateIdle' }),
    )
  })
  it('dispatches attack and damageCharacter actions when isEncounterTurn is true', async () => {
    store.dispatch(damageEncounter(1))

    jest.useFakeTimers()
    store.dispatch = jest.fn()
    render(
      <Provider store={store}>
        <DungeonInBattle />
      </Provider>,
    )

    jest.runAllTimers()
    expect(store.dispatch).toHaveBeenCalledTimes(2)
    expect(store.dispatch).toHaveBeenCalledWith(
      expect.objectContaining({ type: 'game/attack' }),
    )
    expect(store.dispatch).toHaveBeenCalledWith(
      expect.objectContaining({
        type: 'character/damageCharacter',
        payload: expect.any(Number),
      }),
    )
  })
})
