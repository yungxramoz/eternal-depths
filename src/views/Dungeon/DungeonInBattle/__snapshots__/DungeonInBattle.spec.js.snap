// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DungeonInBattle renders correctly: DungeonInBattle 1`] = `
<body>
  <div>
    <div>
      Encounter
    </div>
    <div
      class="progress-container"
    >
      <p
        class="hint-text"
      >
        Your 
         HP 
        100
        /
        100
      </p>
      <div
        class="rpgui-progress"
        data-rpguitype="progress"
      >
        <div
          class=" rpgui-progress-track"
        >
          <div
            class="rpgui-progress-fill red progress-animation"
            style="left: 0px; width: 100%;"
          />
        </div>
        <div
          class=" rpgui-progress-left-edge"
        />
        <div
          class=" rpgui-progress-right-edge"
        />
      </div>
    </div>
    <div
      class="combat-control-container"
    >
      <button
        class="attack-button"
      >
        <div
          class="cooldown-container"
          style="width: 61px; visibility: hidden;"
        />
        <div
          class="rpgui-container container framed framed-golden-2 attack-container"
        >
          <img
            alt="Attack 1"
            class="attack-icon"
            src="test-file-stub"
          />
        </div>
      </button>
    </div>
  </div>
</body>
`;
